src/db/sync/state.rs

```diff
@@ -228,20 +219,68 @@ impl<Db: DatabaseRef + Sync> ConcurrentState<Db> {
         }
     }
 
-    // TODO make cache aware of transitions dropping by having global transition counter.
-    /// Takes the [`BundleState`] changeset from the [`ConcurrentState`],
-    /// replacing it
-    /// with an empty one.
+    /// Create a [`Child`] DB that wraps this state in another cache layer.
     ///
-    /// This will not apply any pending [`TransitionState`]. It is recommended
-    /// to call [`ConcurrentState::merge_transitions`] before taking the bundle.
+    /// This allows the [`ConcurrentState`] to be shared between threads, with
+    /// each thread having a separate cache for its local changes. The child
+    /// can later be merged back into the parent with [`Self::merge_child`] IF
+    /// no other children or copies of the `Arc<Self>` exist.
+    pub fn child(self: &Arc<Self>) -> Child<Db>
+    where
+        Db: Send,
+    {
+        ConcurrentState::new(self.clone(), Default::default())
+    }
+
+    /// Merge a child DB into this DB, incorporating its changes and overwriting
+    /// any present values.
     ///
-    /// If the `State` has been built with the
-    /// [`revm::StateBuilder::with_bundle_prestate`] option, the pre-state will be
-    /// taken along with any changes made by
-    /// [`ConcurrentState::merge_transitions`].
-    pub fn take_bundle(&mut self) -> BundleState {
-        core::mem::take(&mut self.info.bundle_state)
+    /// This function needs to take ownership of the [`Child`], as the child
+    /// copntains an `Arc` of the parent. This [`Arc`] must be dropped to ensure
+    /// that the parent's pointer is unique, allowing the parent to be modified
+    /// via [`Arc::get_mut`].
+    ///
+    /// If other children or other copies of the `Arc<Self>` exist, this will
+    /// fail with a [`ConcurrentStateError`]. error. If the child belongs to
+    /// another parent, this will also fail.
+    ///
+    /// These conditions can be checked via [`Self::can_merge`], please see
+    /// that function's documentation for a possible race condition.
+    pub fn merge_child(self: &mut Arc<Self>, child: Child<Db>) -> Result<(), ConcurrentStateError> {
+        self.can_merge(&child)?;
+
+        let (_, info) = child.into_parts();
+
+        let this = Arc::get_mut(self).ok_or_else(ConcurrentStateError::not_unique)?;
+
+        this.info.cache.absorb(info.cache);
+        Ok(())
+    }
+
+    /// True if the child can be merged into this state, false otherwise.
+    ///
+    /// ## Note
+    ///
+    /// Race conditions can occur if the child is shared between threads, as
+    /// the child may be cloned AFTER this check is made, but BEFORE the child
+    /// is merged. In this case the function will spuriously return `Ok(())`
+    /// even though the merge will fail. To avoid this, ensure that the child
+    /// is not shared between threads when [`Self::can_merge`] is called.
```

Comment:
really really far out idea that might not be worth the trouble, but what if we codify this rule into some state machine-like type

Reply:
would be possible but seems like more effort than it's worth 🤔